### PR TITLE
Use newly implemented polymorphic TAPI models in old implementations

### PR DESCRIPTION
**Pull request by jbartok**
Continues:
* https://github.com/gradle/gradle/pull/29008

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective.
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic.
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`.
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/converters/ConsumerTargetTypeProvider.java [hunk 1]
```diff
@@ -31,28 +31,13 @@
...
 
     @Override
     public <T> Class<? extends T> getTargetType(Class<T> initialTargetType, Object protocolObject) {
-        Class<?>[] interfaces = protocolObject.getClass().getInterfaces();
-        for (Class<?> i : interfaces) {
-            if (configuredTargetTypes.containsKey(i.getName())) {
-                return configuredTargetTypes.get(i.getName()).asSubclass(initialTargetType);
-            }
-        }
+        // TODO: implement this mapping via ToolingModelContract mechanism
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/converters/ConsumerTargetTypeProvider.java:40]: Do we need this todo?

**Review by donat — APPROVED** [review]
Looks good overall! Note, that the cross-version test coverage is executed quite late in the CI pipeline, ie after the `Pull Request` trigger build. I'd suggest to run at least the ready for nightly before merging.

File: platforms/core-runtime/launcher/src/main/java/org/gradle/tooling/internal/provider/serialization/ClientSidePayloadClassLoaderFactory.java [hunk 2]
```diff
@@ -48,90 +35,9 @@ public ClassLoader getClassLoaderFor(ClassLoaderSpec spec, List<? extends ClassL
             if (parents.size() != 1) {
                 throw new IllegalStateException("Expected exactly one parent ClassLoader");
             }
-            return new MixInClassLoader(clSpec.getName() + "-client-payload-loader", parents.get(0), clSpec.getClasspath());
+            return new VisitableURLClassLoader(clSpec.getName() + "-client-payload-loader", parents.get(0), clSpec.getClasspath());
         }
         return classLoaderFactory.getClassLoaderFor(spec, parents);
     }
 
-    private static class MixInClassLoader extends TransformingClassLoader {
```
Comment (adammurdoch) [inline platforms/core-runtime/launcher/src/main/java/org/gradle/tooling/internal/provider/serialization/ClientSidePayloadClassLoaderFactory.java:56]: 👍

**Review by adammurdoch — APPROVED** [review]

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test RFR

**Comment by jbartok** [comment]
@bot-gradle test RFR

**Comment by jbartok** [comment]
Bad news. I think this PR is a dud... We're trying to replace apples with oranges, unfortunately.

[What we have implemented for polymorphism](https://github.com/gradle/gradle/pull/29008) is a mechanism that takes one specific model implementation and when building a proxy from it, mimics multiple interfaces implemented by that specific model implementation.

But what happens in the case of [DefaultIdeaModuleDependency](https://github.com/gradle/gradle/blob/abb4749a2162492f8938320f8927ae4234f14900/platforms/ide/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/idea/DefaultIdeaModuleDependency.java#L22) is that we actually [generate bytecode](https://github.com/gradle/gradle/blob/598e8dfbbfa7b3c5822d0cf210a19c53d309b04f/platforms/core-runtime/client-services/src/main/java/org/gradle/internal/daemon/client/serialization/ClientSidePayloadClassLoaderFactory.java#L70) for an [old method in some old interface](https://github.com/gradle/gradle/blob/866060514f01806785b1aef1c029dc402175f41a/platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/converters/BackwardsCompatibleIdeaModuleDependency.java#L26), a method which is not in the model implementation anymore.

And this is [just one of the magics done by the MixinClassLoader](https://github.com/gradle/gradle/blob/866060514f01806785b1aef1c029dc402175f41a/platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/HasCompatibilityMapping.java#L45), unfortunately.

MAYBE we could get rid of the IdeaModuleDependency - BackwardsCompatibleIdeaModuleDependency magic when we reach Gradle 9 and (if?) we raise the TAPI support lower limit higher than Gradle 3.0, (that was the version failing in tests, which made me dig deeper), but I don't even know what all the other mixins do...

Wdyt?

**Pull request closed without merging**